### PR TITLE
fix #76: Safe exception messages provide argument data

### LINKED ISSUE #76: Safe Exception Implementations should provide arguments in message
ServiceException, for example, provides an unsafe exception message (`getMessage()`), and a safe log message (`getLogMessage()`). By providing both we can avoid losing data when libraries which depend on safe-logging are used in environments without sls-logging appenders.

Current message when the following is logged using non-sls mechanisms `new SafeIllegalArgumentException("Bad Value", SafeArg.of("value", 3))`

`Bad Value`

Expected:

`Bad Value: {value=3}`

### PR DESCRIPTION
**Pull request by carterkozak**
Only logMessage is redacted.

### CODE REVIEW

**Comment by carterkozak** [comment]
fyi @uschi2000 @iamdanfox

File: preconditions/src/main/java/com/palantir/logsafe/exceptions/SafeExceptions.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            Arg<?> argument = args[i];
+            sb.append(argument.getName())
+                    .append("=")
+                    .append(argument.getValue());
+            if (i < args.length - 1) {
+                sb.append(", ");
+            }
+        }
+        return sb.append('}')
+                .toString();
```
Comment (dansanduleac) [inline preconditions/src/main/java/com/palantir/logsafe/exceptions/SafeExceptions.java]: one line

File: preconditions/src/main/java/com/palantir/logsafe/exceptions/SafeExceptions.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+final class SafeExceptions {
+    private SafeExceptions() {}
+
+    static String renderMessage(String message, Arg<?>... args) {
+        StringBuilder sb = new StringBuilder(message).append(": {");
+        for (int i = 0; i < args.length; i++) {
+            Arg<?> argument = args[i];
+            sb.append(argument.getName())
+                    .append("=")
+                    .append(argument.getValue());
```
Comment (dansanduleac) [inline preconditions/src/main/java/com/palantir/logsafe/exceptions/SafeExceptions.java]: one line

File: preconditions/src/main/java/com/palantir/logsafe/exceptions/SafeExceptions.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            if (i < args.length - 1) {
+                sb.append(", ");
+            }
```
Comment (dansanduleac) [inline preconditions/src/main/java/com/palantir/logsafe/exceptions/SafeExceptions.java:38]: maybe we should just copy this code from conjure-java-runtime which already does what you want!
https://github.com/palantir/conjure-java-runtime-api/blob/8d23fc7fb11481033ca4867fdb99a93186b108ac/errors/src/main/java/com/palantir/conjure/java/api/errors/ServiceException.java#L95-L117

Comment (dansanduleac) [reply preconditions/src/main/java/com/palantir/logsafe/exceptions/SafeExceptions.java:38]: (except the first line, but e.g. it avoids printing out `: {}` in the trivial case where there are no arguments)

**Comment by carterkozak** [comment]
Updated, thanks for the feedback

File: preconditions/src/main/java/com/palantir/logsafe/exceptions/SafeExceptions.java (hunk 3 above)
Comment (carterkozak) [reply preconditions/src/main/java/com/palantir/logsafe/exceptions/SafeExceptions.java:38]: done

**Review by dansanduleac — APPROVED** [review]

**Comment by iamdanfox** [comment]
@dansanduleac @cakofony I think this PR is not entirely safe because people may have assumed that the message of a `SafeRuntimeException` is safe to log.  

Before this PR, that would have been true because we have [baseline-error-prone](https://github.com/palantir/gradle-baseline/blob/develop/baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/LogSafePreconditionsMessageFormat.java) enforcing that people only write constant messages.

However after this PR, the following snippet could cause a leak because `e.getMessage()` contains the unsafe arg 'danger':

```java
try {
  throw new SafeRuntimeException("...", UnsafeArg("danger", danger));
  // ...

catch (SafeRuntimeException e) {
	log.info("Ignoring something and retrying", SafeArg.of("message", e.getMessage());
	//                                                                ^ 🔥🔥🔥
    retry();
}
```

As a compromise, could we just omit all the UnsafeArgs from the exception message and accept that they will just never be visible in logs produced by non-palantir slf4j implementations?  Alternatively, I think we need to urgently extend our errorprone checks to prevent the above scenario.

**Comment by carterkozak** [comment]
Only including safe args doesn’t provide enough data to debug applications which do not produce sls service logs.

Logging a safe arg of getMessage is strictly worse than using getLogMessage, since only the latter is guaranteed to be safe.

**Comment by iamdanfox** [comment]
Any chance you could pop an extra errorprone check into baseline then? I think this means we could wholesale ban anyone writing `SafeArg.of("...", e.getMessage())`

**Comment by carterkozak** [comment]
Sure thing

**Comment by carterkozak** [comment]
@iamdanfox Cheers https://github.com/palantir/gradle-baseline/pull/444
We may also want this in order to differentiate between throwable messages and SafeLoggable log messages in testing: https://github.com/palantir/safe-logging/pull/78

**Pull request merged**